### PR TITLE
102 redactor newline corruption clean

### PR DESCRIPTION
**Pull request by bennyyang11**
# Fix: Preserve file integrity in redaction pipeline

## Summary

Fixed redactors unconditionally adding newlines, which corrupted binary files during support bundle collection.

**Before:** Binary files grew from 51 → 53 bytes (extra `0x0a 0x0a` appended)  
**After:** Binary files preserved byte-for-byte, SHA256 checksums match

[Issue #102](https://github.com/replicated-collab/pixee-replicated/issues/102)

## Problem

The redaction pipeline was corrupting binary files by adding newlines where none existed:

- Copy collector copies file from container →  Clean (51 bytes)
- Redaction processes file →  Corrupted (53 bytes, +2 newlines)
- Result: Binary files unusable, checksums don't match

**Root cause:** `bufio.Scanner` and `ReadLine()` strip newlines during read, then the code unconditionally added them back, assuming all files were text with newline terminators.

## Solution

Created `LineReader` utility that tracks whether each line originally had a newline:
- Only restores newlines when they were present in the original
- Refactored SingleLineRedactor, MultiLineRedactor, and LiteralRedactor
- Preserves byte-for-byte integrity for all file types

## Changes

- `pkg/redact/line_reader.go` (NEW): Newline-aware line reader
- `pkg/redact/single_line.go`: Conditional newline restoration  
- `pkg/redact/multi_line.go`: Fixed EOF handler, conditional newlines
- `pkg/redact/literal.go`: Conditional newline restoration
- Added 48 comprehensive tests

## Testing

- [x] All existing tests pass
- [x] 48 new tests covering binary files, text files, edge cases  
- [x] Manual verification: 51-byte binary → 51 bytes (not 53)  
- [x] SHA256 checksums match source files  
- [x] Copy collector → redaction flow verified

## Backwards Compatibility

Support bundle checksums will change for files that previously had newlines incorrectly added. This is correct behavior - files should match their source. The old behavior was corrupting files.

## Checklist

- [x] New and existing tests pass locally
- [x] Tests for the changes have been added
- [x] Commit messages are informative
- [x] Documentation added/updated

## Breaking Change?

- [ ] Yes
- [x] No

*Note: Checksum changes are a bug fix, not a breaking change.*

### CODE REVIEW

**Review by hedge-sparrow — DISMISSED** [review]
lgtm

File: pkg/collect/collector_test.go [hunk 1]
```diff
@@ -249,8 +243,7 @@ abc`,
 abc
 123
 xyz123
-abc
-`,
```
Comment (laverya) [inline pkg/collect/collector_test.go:253]: I would love a test that looks at different types of line feed characters - do we accidentally do `\r\n`->`\n` conversions for instance?

File: pkg/redact/literal.go [hunk 2]
```diff
@@ -34,7 +39,8 @@ func (r literalRedactor) Redact(input io.Reader, path string) io.Reader {
 			if err == nil || err == io.EOF {
 				writer.Close()
 			} else {
-				if err == bufio.ErrTooLong {
+				// Check if error is about line exceeding maximum size
+				if err != nil && bytes.Contains([]byte(err.Error()), []byte("exceeds maximum size")) {
```
Comment (laverya) [inline pkg/redact/literal.go]: can we return a typed error to avoid having to parse this, or even return bufio.ErrTooLong directly?

File: pkg/redact/literal_test.go [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+	binaryData := []byte{0x01, 0x02, 0x03, 0x04, 0x00, 0xFF, 0xFE, 0xAB, 0xCD}
+
+	redactor := literalString([]byte("notfound"), "testfile", t.Name())
+
+	out := redactor.Redact(bytes.NewReader(binaryData), "test.bin")
+	result, err := io.ReadAll(out)
+
+	require.NoError(t, err)
+	require.Equal(t, binaryData, result, "Binary file should be unchanged")
+}
```
Comment (laverya) [inline pkg/redact/literal_test.go:83]: Can you add a binary test with every single byte value? 0x01 -> 0xFF

Comment (laverya) [reply pkg/redact/literal_test.go:83]: maybe even every two byte combination, 0x00+0x00 -> 0xFF+0xFF
only 64k

File: pkg/redact/multi_line_test.go [hunk 4]
```diff
@@ -158,3 +152,224 @@ func Test_writeBytes(t *testing.T) {
...
 	}
 }
+
+// Test 3.16: Binary file (no newlines) → unchanged
+func TestMultiLineRedactor_BinaryFile(t *testing.T) {
+	ResetRedactionList()
+	defer ResetRedactionList()
+
+	// Binary content with no newlines - the bug that caused 2 extra bytes
+	binaryData := []byte{0x01, 0x02, 0x03, 0x04, 0x00, 0xFF, 0xFE, 0xAB, 0xCD}
```
Comment (laverya) [inline pkg/redact/multi_line_test.go:162]: I want a 64k version of this test too, just like with literal 😄 

(keep this one of course!)

File: pkg/redact/single_line_test.go [hunk 5]
```diff
@@ -340,3 +334,214 @@ func TestNewSingleLineRedactor(t *testing.T) {
...
 }
+
+// Test 2.15: Binary file (no newlines) → unchanged
+func TestSingleLineRedactor_BinaryFile(t *testing.T) {
+	ResetRedactionList()
+	defer ResetRedactionList()
+
+	// Binary content with no newlines
+	binaryData := []byte{0x01, 0x02, 0x03, 0x04, 0x00, 0xFF, 0xFE, 0xAB, 0xCD}
+
```
Comment (laverya) [inline pkg/redact/single_line_test.go:345]: great test, I want the 64k test here too

**Review by laverya — APPROVED** [review]
